File: electron/main/ipc/library.ts
import { ipcMain } from 'electron';
import type {
	LibraryAlbumArtistKey,
	LibraryAlbumArtistSortKey,
	LibraryAlbumSortKey,
	LibraryArtistAlbumSortKey,
	LibrarySortDirection,
	LibraryTrackSortKey
} from '@shared/native-backend';
import type { NativeBackend } from '@shared/native-backend';
import { validateSender } from '../security';
import {
	requireAlbumKey,
	requireBoolean,
	requireNullableString,
	requireOneOf,
	requireString
} from './validation';

const SORT_DIRECTIONS = [
	'ascending',
	'descending'
] as const satisfies readonly LibrarySortDirection[];
const TRACK_SORT_KEYS = [
	'title',
	'artist',
	'album',
	'duration'
] as const satisfies readonly LibraryTrackSortKey[];
const ALBUM_SORT_KEYS = [
	'title',
	'artist',
	'year'
] as const satisfies readonly LibraryAlbumSortKey[];
const ARTIST_SORT_KEYS = [
	'artist',
	'albumCount',
	'trackCount'
] as const satisfies readonly LibraryAlbumArtistSortKey[];
const ARTIST_ALBUM_SORT_KEYS = [
	'year',
	'title'
] as const satisfies readonly LibraryArtistAlbumSortKey[];

function requireArtistKey(value: unknown): LibraryAlbumArtistKey {
	if (
		typeof value !== 'object' ||
		value === null ||
		typeof (value as { name?: unknown }).name !== 'string'
	)
		throw new TypeError('artistKey must contain a name string');
	return value as LibraryAlbumArtistKey;
}

export function registerLibraryIpc(backend: NativeBackend): void {
	ipcMain.handle(
		'library:get-status',
		validateSender(() => backend.getLibraryStatus())
	);
	ipcMain.handle(
		'library:get-track-for-path',
		validateSender((path: unknown) => backend.getLibraryTrackForPath(requireString(path, 'path')))
	);
	ipcMain.handle(
		'library:list-roots',
		validateSender(() => backend.listLibraryRoots())
	);
	ipcMain.handle(
		'library:register-root',
		validateSender((path: unknown) => backend.registerLibraryRoot(requireString(path, 'path')))
	);
	ipcMain.handle(
		'library:set-root-enabled',
		validateSender((id: unknown, enabled: unknown) =>
			backend.setLibraryRootEnabled(requireString(id, 'id'), requireBoolean(enabled, 'enabled'))
		)
	);
	ipcMain.handle(
		'library:remove-root',
		validateSender((id: unknown) => backend.removeLibraryRoot(requireString(id, 'id')))
	);
	ipcMain.handle(
		'library:get-scan-state',
		validateSender(() => backend.getLibraryScanState())
	);
	ipcMain.handle(
		'library:start-scan',
		validateSender(() => backend.startLibraryScan())
	);
	ipcMain.handle(
		'library:cancel-scan',
		validateSender(() => backend.cancelLibraryScan())
	);
	ipcMain.handle(
		'library:list-tracks',
		validateSender((cursor: unknown, search: unknown, sortKey: unknown, sortDirection: unknown) =>
			backend.listLibraryTracks(
				requireNullableString(cursor, 'cursor'),
				requireNullableString(search, 'search'),
				requireOneOf(sortKey, 'sortKey', TRACK_SORT_KEYS),
				requireOneOf(sortDirection, 'sortDirection', SORT_DIRECTIONS)
			)
		)
	);
	ipcMain.handle(
		'library:list-albums',
		validateSender((cursor: unknown, search: unknown, sortKey: unknown, sortDirection: unknown) =>
			backend.listLibraryAlbums(
				requireNullableString(cursor, 'cursor'),
				requireNullableString(search, 'search'),
				requireOneOf(sortKey, 'sortKey', ALBUM_SORT_KEYS),
				requireOneOf(sortDirection, 'sortDirection', SORT_DIRECTIONS)
			)
		)
	);
	ipcMain.handle(
		'library:list-album-artists',
		validateSender((cursor: unknown, search: unknown, sortKey: unknown, sortDirection: unknown) =>
			backend.listLibraryAlbumArtists(
				requireNullableString(cursor, 'cursor'),
				requireNullableString(search, 'search'),
				requireOneOf(sortKey, 'sortKey', ARTIST_SORT_KEYS),
				requireOneOf(sortDirection, 'sortDirection', SORT_DIRECTIONS)
			)
		)
	);
	ipcMain.handle(
		'library:get-album-artist',
		validateSender((artistKey: unknown) =>
			backend.getLibraryAlbumArtist(requireArtistKey(artistKey))
		)
	);
	ipcMain.handle(
		'library:list-artist-albums',
		validateSender(
			(artistKey: unknown, cursor: unknown, sortKey: unknown, sortDirection: unknown) =>
				backend.listLibraryArtistAlbums(
					requireArtistKey(artistKey),
					requireNullableString(cursor, 'cursor'),
					requireOneOf(sortKey, 'sortKey', ARTIST_ALBUM_SORT_KEYS),
					requireOneOf(sortDirection, 'sortDirection', SORT_DIRECTIONS)
				)
		)
	);
	ipcMain.handle(
		'library:get-album-details',
		validateSender((albumKey: unknown) => backend.getLibraryAlbumDetails(requireAlbumKey(albumKey)))
	);
	ipcMain.handle(
		'library:list-album-tracks',
		validateSender((albumKey: unknown, cursor: unknown) =>
			backend.listLibraryAlbumTracks(
				requireAlbumKey(albumKey),
				requireNullableString(cursor, 'cursor')
			)
		)
	);
	ipcMain.handle(
		'library:start-track',
		validateSender((trackId: unknown) =>
			backend.startLibraryTrack(requireString(trackId, 'trackId'))
		)
	);
	ipcMain.handle(
		'library:start-album',
		validateSender((albumKey: unknown) => backend.startLibraryAlbum(requireAlbumKey(albumKey)))
	);
}

File: src-tauri/src/commands/library.rs
use backend::{
    audio::playback::PlaybackSnapshot,
    library::models::{
        LibraryAlbumArtistKey, LibraryAlbumArtistSortKey, LibraryAlbumKey, LibraryAlbumSortKey,
        LibraryArtistAlbumSortKey, LibrarySortDirection, LibraryTrackSortKey,
    },
    library::service::{LibraryCommandError, StartLibraryAlbumTrackError, StartLibraryTrackError},
};

use crate::AppState;

#[tauri::command]
#[specta::specta]
pub fn get_library_status(
    state: tauri::State<'_, AppState>,
) -> backend::library::models::LibraryStatus {
    state.backend.library.status()
}

#[tauri::command]
#[specta::specta]
pub fn get_library_scan_state(
    state: tauri::State<'_, AppState>,
) -> backend::library::models::LibraryScanSnapshot {
    state.backend.library.handle().scan_state()
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_roots(
    state: tauri::State<'_, AppState>,
) -> Result<Vec<backend::library::models::LibraryRoot>, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.roots())
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn register_library_root(
    path: String,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryRoot, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.register_root(path))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn set_library_root_enabled(
    id: String,
    enabled: bool,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryRoot, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.set_root_enabled(id, enabled))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub fn remove_library_root(
    id: String,
    state: tauri::State<'_, AppState>,
) -> Result<(), LibraryCommandError> {
    state.backend.library.handle().remove_root(id)
}

#[tauri::command]
#[specta::specta]
pub fn start_library_scan(state: tauri::State<'_, AppState>) -> Result<(), LibraryCommandError> {
    state.backend.library.handle().start_scan()
}

#[tauri::command]
#[specta::specta]
pub fn cancel_library_scan(state: tauri::State<'_, AppState>) -> Result<(), LibraryCommandError> {
    state.backend.library.handle().cancel_scan()
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_tracks(
    cursor: Option<String>,
    search: Option<String>,
    sort_key: LibraryTrackSortKey,
    sort_direction: LibrarySortDirection,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryTrackPage, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || {
        library.catalog_tracks(cursor, search, sort_key, sort_direction)
    })
    .await
    .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_albums(
    cursor: Option<String>,
    search: Option<String>,
    sort_key: LibraryAlbumSortKey,
    sort_direction: LibrarySortDirection,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumPage, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || {
        library.catalog_albums(cursor, search, sort_key, sort_direction)
    })
    .await
    .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_album_artists(
    cursor: Option<String>,
    search: Option<String>,
    sort_key: LibraryAlbumArtistSortKey,
    sort_direction: LibrarySortDirection,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumArtistPage, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || {
        library.catalog_album_artists(cursor, search, sort_key, sort_direction)
    })
    .await
    .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn get_library_album_artist(
    artist_key: LibraryAlbumArtistKey,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumArtistSummary, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.catalog_artist(artist_key))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_artist_albums(
    artist_key: LibraryAlbumArtistKey,
    cursor: Option<String>,
    sort_key: LibraryArtistAlbumSortKey,
    sort_direction: LibrarySortDirection,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumPage, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || {
        library.catalog_artist_albums(artist_key, cursor, sort_key, sort_direction)
    })
    .await
    .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn get_library_album_details(
    album_key: LibraryAlbumKey,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumDetails, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.catalog_album_details(album_key))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn list_library_album_tracks(
    album_key: LibraryAlbumKey,
    cursor: Option<String>,
    state: tauri::State<'_, AppState>,
) -> Result<backend::library::models::LibraryAlbumTrackPage, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.catalog_album_tracks(album_key, cursor))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn get_library_track_for_path(
    path: String,
    state: tauri::State<'_, AppState>,
) -> Result<Option<backend::library::models::LibraryTrackSummary>, LibraryCommandError> {
    let library = state.backend.library.handle();
    tauri::async_runtime::spawn_blocking(move || library.track_for_path(path))
        .await
        .map_err(|_| LibraryCommandError::TaskFailed)?
}

#[tauri::command]
#[specta::specta]
pub async fn start_library_track(
    track_id: String,
    state: tauri::State<'_, AppState>,
) -> Result<PlaybackSnapshot, StartLibraryTrackError> {
    state.backend.start_library_track(track_id).await
}

#[tauri::command]
#[specta::specta]
pub async fn start_library_album(
    album_key: LibraryAlbumKey,
    state: tauri::State<'_, AppState>,
) -> Result<PlaybackSnapshot, StartLibraryAlbumTrackError> {
    state.backend.start_library_album(album_key).await
}
